Refactor Tests.

amanagr opened the pull request.

neiljp commented on tests/conftest.py:
```diff
@@ -100,6 +100,43 @@ def users_fixture():
+
+@pytest.fixture(scope='module')
+def streams_fixture():
+    streams = [denmark, secret_stream]
```
Is there a reason why you included denmark here?

amanagr replied:
Yeah, I wanted to separate out the unique example for each fixtures. Like general stream, private stream etc. It will be helpful later if you just want to use one type of fixture in your test.

neiljp commented on tests/conftest.py:
```diff
@@ -590,9 +546,10 @@ def streams():
     `initial_data` fixture.
     """
     return [
```
Should this be calculated?

amanagr replied:
Yeah totally.

neiljp reviewed (commented):
@amanagr I think this will be a lot clearer and help people understand our tests much better, and also broaden our coverage, by not being so verbose :+1:

* As per comments, can we use logged_on_user with the new fixture?

* For the initial_data:
  - I think it's important to have (at least some of) the specific cross-realm bots which are expected, perhaps as a separate fixture. For just 'bot', I was expecting a user-derived bot, not a system-bot.
  - are we not missing some metadata which you recorded from the server? Maybe use a list comprehension for that, from the minimal users data?

I like the direction this is heading :)

neiljp commented:
@sumanthvrao Some of this follows your ideas in #374, so it would be good to hear your thoughts on this.

amanagr commented on tests/conftest.py:
```diff
@@ -74,6 +74,33 @@ def msg_box(mocker, messages_successful_response):
+            'full_name': 'Human {}'.format(i),
+            'email': 'person{}@example.com'.format(i),
+        })
+    return deepcopy(users)
```
If we don't use module scope we won't we able to use it in fixtures which are module-scope.

amanagr commented on tests/conftest.py:
```diff
@@ -236,14 +226,14 @@ def messages_successful_response() -> Dict[str, Any]:
     """
     Response from /register API request.
     """
     return {
```
Yeah, It's better to remove module scope for initial_data fixture. Cool! I will transform me into logged_on_user.

amanagr replied:
We will still have to use deepcopy inside users_fixture for it's use in future functions which might be module scoped.

sumanthvrao commented on tests/model/test_model.py:
```diff
@@ -481,6 +481,7 @@ def test__stream_info_from_subscriptions(self, initial_data, streams,
         assert all(msg_id == msg['stream_id'] for msg_id, msg in by_id.items())
         assert muted_streams == muted
         assert pinned == []  # FIXME generalize/parametrize
+        print(unpinned)
```
@amanagr pinging you if you missed this :)

sumanthvrao replied:
The 2 files here, `test_model` and `test_ui_tools` here are adjustments due to the previous commits introduced right? Maybe we can move them to a new commit (Just a suggestion).

sumanthvrao reviewed (commented):
These changes simplify our tests a great deal :+1: . This was also what I was going for with #374;

* I echo what Neil said about the bots fixture (which we could introduce). Maybe we can have a user and a cross-realm bot ass 2 parameters in it?

Left 2 small suggestions/ comments about this. Hope it helps :)

amanagr commented on tests/conftest.py:
```diff
@@ -103,6 +103,43 @@ def logged_on_user():
         'short_name': 'Human',
     }
 
+denmark = {
```
Cool, sounds better!

neiljp commented on tests/conftest.py:
```diff
@@ -264,78 +264,8 @@ def initial_data(logged_on_user):
-                          '78eecc367eedd27e6ac9292dc966beb6'
-                          '?d=identicon&version=1'
-        }],
+        'realm_users': users_fixture,
```
The data in `users_fixture` doesn't seem to match the extent we have currently?

neiljp reviewed (commented):
@amanagr I like the way this is turning out :+1: I think we need to be careful we don't lose any coverage though, and some of the data structures look different in various ways?

I can see that module scope for fixtures might make sense, but my feeling is that it has specific purposes, and in most cases having them with regular scope shouldn't make a big speed difference, but does ensure that data doesn't get mixed between tests by mistake - which seems like a more important factor? I've not examined if there's a speed difference, though.

Naming-wise, I'm wondering if it would be clearer to use a prefix for the source/target of the fixture data, eg. `server__initial_data` when the fixture represents json from a specific server call, or `model__user_dict` (as opposed to the proposed `users_dict`) for the expected `user_dict` to generate from the former? There are other intermediate fixtures to name too, but what do you think of that kind of idea? Then it becomes clearer if we're using a server response in a test as input, vs comparing it to other fixtures?

I've not reviewed the last commit yet, as it's a new test, rather than refactoring/rewriting the existing tests.

amanagr commented:
For streams and users I have tried to trim to the data we are not using in tests. Do we really need all the data for testing? If we add a feature in the future that uses some data not available in the fixtures, then we can simply add that.

For changes within tests, I am not sure what would be the best way to show it as I have changed the stream/user id's in conftest. The id changes are required since in future if we can easily generate say 10 streams by increasing the range in the streams_fixtures or giving `n` as no. of streams to return.

Do you want me to squash the `Add` and `Use` fixtures?

neiljp commented:
Re 'all the data', I'm not suggesting detailed specific users like we had before - I like the general ones. However, my understanding is that the cross-realm bots are intrinsic to the server (typically, at least). Also, the subscriptions data seems incomplete compared to before - this was why I was trying to suggest having new fixtures with removal of old hard-coded data and use of the new fixtures in the same commit, to show what code is substituted.

If squashing those two commits looks reasonable, then I'd say give it a try (you can always keep a copy).

(The fixture naming is less important, but I thought it might be useful to see the context)

amanagr commented:
Cool! I understand now.

neiljp reviewed (commented):
@amanagr This iteration is a lot easier to follow, commit-wise :+1:

Did you notice much difference in performance with the different scope? It seems simpler now?

amanagr commented on tests/conftest.py:
```diff
@@ -100,6 +100,62 @@ def logged_on_user():
+            'stream_id': i,
+            'in_home_view': True,
+            'audible_notifications': False,
+            'description': 'Stream',
```
Cool! Done.

amanagr commented on tests/conftest.py:
```diff
@@ -348,10 +360,26 @@ def initial_data(logged_on_user, users_fixture):
+                'topic': 'Denmark2',
+                'unread_message_ids': [7],
+                'sender_ids': [1, 2]
+            }],
```
Yeah! :/

amanagr commented on tests/conftest.py:
```diff
@@ -612,16 +640,15 @@ def user_list(logged_on_user):
+    return [['Secret stream', 99, '#c3c3c3', True],
+            ['Some general stream', 1000, '#b0a5fd', False],
+            ['Stream 1', 1, '#b0a5fd', False],
+            ['Stream 2', 2, '#b0a5fd', False]]
```
No, since the list's order is important, I think we should leave this as it is.

amanagr commented on tests/conftest.py:
```diff
@@ -542,47 +533,47 @@ def user_dict(logged_on_user):
-            'full_name': 'Welcome Bot',
-            'status': 'inactive',
-            'user_id': 4
+        logged_on_user['email']: {
```
No, I moved it up.

amanagr commented on tests/conftest.py:
```diff
@@ -74,6 +74,23 @@ def msg_box(mocker, messages_successful_response):
+    users = [logged_on_user]
+    for i in range(1, 3):
+        users.append({
+            'user_id': i,
```
Yeah, nice.

neiljp reviewed (commented):
@amanagr I'm really looking forward to having this new version of the tests :+1:

The last commit seems a little different, so I've mainly focused on the refactoring and ensuring we don't lose anything in the transformation, though the last commit seems pretty simple in terms of the additional test.

amanagr commented:
@neiljp updated! Thanks for the ideas :)

neiljp commented:
@amanagr To keep this moving and make it easier to manage I've pushed three commits to master now.

Is the `avatar_url` consistent? Is it now absent due to the client_gravatar setting?

I was going to merge the streams_fixture commit too, but it is causing some tests to fail for me, and the tests were failing generally too on this PR, so that may be related.

I've not looked at the time-based commit; in principle I like the idea, but it makes the tests less reproducible, so I would be happy using a 'seed' value instead, a little like we used to have?

amanagr commented:
Yeah, the `avatar_url` is not required since we will never use it, it's just dead data. I will try and fix the tests, strangely the tests passed last time I pushed changes and now without any changes made by me they have severely broken!

neiljp commented:
@amanagr Re the avatar_url, I was interested in whether this is data which we now don't receive, or which we just don't process. If the former, there is zero issue in just removing it; if the latter, I'm inclined to put a 'placeholder' in there, maybe even with a comment on each line as to why it's still there? In other cases we have kept lots of fields (eg for users or streams?) which we are not currently using. I'm not suggesting we'll use the avatar images in the foreseeable future - though we could! - but rather that while we've simplified/abstracted the fixtures received from the server, they should still map to the data we receive.

One possibility that I thought of while looking at this, would be a tool that sets up a 'fresh' dev install in a certain way, and then fetches candidate fixture elements that can be compared to what we have. That would allow tracking across API versions, extra fields, etc., and ensuring we don't lose any data that may (or may not) be there. That *also* sounds like the thoughts from before of having a 'template dev system' to get automated screenshots from, but let's not dive into that too fast yet :)

amanagr committed:
conftest: Cleanup long avatar_url and use dummy url instead.

neiljp reviewed (commented):
@amanagr I merged another 4 of these almost as-is, with some extra descriptive commit text. The cleanliness of the tests looks so much nicer now :)

The remaining issues are minor, and another was a follow-up on an earlier comment in terms of us 'losing' some test fixtures that had different presence states, which can go into another PR (or into an issue).

I also think we should check through and see if there are older stream ids that we should strip out (86, 14)? Of course, also any other further refactoring or extensions, now that things are cleaner :)

The classify counts I merged: I assume you can put a simple model-related test/code PR on that related to group-PMs/huddles? Also, should we have a straight 'nothing muted' case, do you think?

amanagr committed:
conftest: Add offlline and idle type users to user list.

This improves testing of get_all_users too.
Add fixture to fix current time which makes time tests consistent.

amanagr commented on tests/conftest.py:
```diff
@@ -352,9 +346,7 @@ def initial_data(logged_on_user, users_fixture, streams_fixture):
-            'avatar_url': 'https://secure.gravatar.com/avatar/'
-                          '78eecc367eedd27e6ac9292dc966beb6'
-                          '?d=identicon&version=1'
+            'avatar_url': 'avatar_url'
```
Done!

amanagr commented on tests/conftest.py:
```diff
@@ -368,12 +370,31 @@ def initial_data(logged_on_user, users_fixture):
+                    'pushable': False,
+                    'client': 'ZulipMobile',
+                    'status': 'idle',
+                    'timestamp': 1522384165
```
I have made them static for now, We can work on timestamps like `'timestamp': 1522384165` in a different PR. I have added it to my TODO list for now.

neiljp replied:
You still have `time.time()`, which doesn't seem very static?

amanagr replied:
There is `mock_current_time` in parameters of the function which makes it static.

neiljp replied:
I agree it keeps it static, but I'm unclear why you're using a real (dynamic) function in test code, but then mock it.

amanagr commented on tests/conftest.py:
```diff
@@ -548,13 +569,13 @@ def user_dict(logged_on_user):
             'email': 'person2@example.com',
             'user_id': 12,
-            'status': 'inactive'
+            'status': 'idle'
```
Added to TODO list.

neiljp commented:
@amanagr I realized we hadn't merged these;  I've pushed the avatar_url with an extra note in the summary, but could we skip what seems to be unnecessary mocking in the other commit?
